### PR TITLE
[MRG+1] CDU distributive relations

### PR DESCRIPTION
**Pull request by jrmyp**
This PR adds a new way to strip CDUs from a graph.

So far relation sources/targets are relocated to the recursive head of the CDUs when stripping. The new modes trigger distribution over all components of the CDUs, always (for `broadcast` mode) or depending on the label (`custom` mode). The process creates new edges.

I'm working on attelo integration of the process, and I will probably add new commits.

### CODE REVIEW

**Comment by moreymat** [comment]
Thanks for the PR @jrmyp !
This looks good, shallowly, but I will have a thorough look at it tomorrow morning with a clear mind.

I would love to have more frequent and numerous PRs !

**Comment by jrmyp** [comment]
More PRs will definitely come. The last weeks haven't been very productive on my part. It's better now.

I had forgotten that attelo doesn't drive the feature extraction (everything is made from educe), so the integration I'm doing will in fact be part of this PR.

**Comment by moreymat** [comment]
@jrmyp I hope you have seen PR #79 (merged into master 2.5 weeks ago) for updated statistics on the STAC corpus, including much more detailed statistics on CDUs and relations.

File: educe/graph.py [hunk 1]
```diff
@@ -502,16 +502,15 @@ def cdu_members(self, cdu, deep=False):
...
         """
 
+        hyperedge = self.edgeform(cdu)
+        members = set(self.links(hyperedge))
+
         if deep:
-            members = set()
-            for m in self.cdu_members(cdu):
-                members.add(m)
+            for m in list(members):
```
Comment (moreymat) [inline educe/graph.py]: You can already iterate over the set form of `members`, so why convert into a list?
If this has a noticeable impact on speed for our typical use cases, we can keep this but you should add a comment stating that this conversion to list is just an optimization (as opposed to a semantically motivated conversion).

File: educe/graph.py [hunk 2]
```diff
@@ -502,16 +502,15 @@ def cdu_members(self, cdu, deep=False):
...
                 if self.is_cdu(m):
-                    members.update(self.cdu_members(m,deep))
-            return frozenset(members)
-        else:
-            hyperedge = self.edgeform(cdu)
-            return frozenset(self.links(hyperedge))
+                    members.update(self.cdu_members(m, True))
```
Comment (moreymat) [inline educe/graph.py]: I think `self.cdu_members(m, deep=deep)` should be preferred.

File: educe/stac/graph.py [hunk 3]
```diff
@@ -148,7 +139,11 @@ def without_cdus(self, sloppy=False):
         just as easily call deepcopy yourself
         """
         g2 = copy.deepcopy(self)
-        g2.strip_cdus(sloppy)
+        if mode == 'head':
+            g2.strip_cdus(sloppy)
```
Comment (moreymat) [inline educe/stac/graph.py]: Same as above for named parameters: `strip_cdus(sloppy=sloppy)`.

File: educe/stac/graph.py [hunk 4]
```diff
@@ -148,7 +139,11 @@ def without_cdus(self, sloppy=False):
...
+        elif mode == 'broadcast' or mode == 'custom':
+            g2.strip_cdus_distribute(sloppy, mode)
```
Comment (moreymat) [inline educe/stac/graph.py]: Ibid. for named parameters.

File: educe/stac/graph.py [hunk 5]
```diff
@@ -197,6 +192,92 @@ def strip_cdus(self, sloppy=False):
...
+                    # Either distribute over all components...
+                    nodes = edu_components(node)
+                else:
+                    # ... or link to the CDU recursive head only
+                    nodes = [heads[self.mirror(node)]]
+                return nodes
+
+            return [candidates(node, lset) for (node, lset) in
+                    ((src_node, LEFT_DIST), (tgt_node, RIGHT_DIST))]
+
```
Comment (moreymat) [inline educe/stac/graph.py]: The list comprehension seems superfluous and miscarries the intended return type (a pair).
I would prefer the following:
```
return (candidates(src_node, LEFT_DIST),
            candidates(tgt_node, RIGHT_DIST))
```

File: educe/stac/graph.py [hunk 6]
```diff
@@ -197,6 +192,92 @@ def strip_cdus(self, sloppy=False):
...
+            """ Returns a list of all EDUs contained by a node. """
+            if self.is_edu(node):
+                return [node]
+            return [snode for snode in self.cdu_members(node, deep=True)
+                        if self.is_edu(snode)]
+
+        # Convert all edges in order
+        for old_edge in self.relations():
+            links = self.links(old_edge)
+            assert(len(links) == 2) # Verify the edge is well-formed
```
Comment (moreymat) [inline educe/stac/graph.py]: This line has somehow escaped your pep8 checker, it seems :-)

File: educe/stac/graph.py [hunk 7]
```diff
@@ -118,27 +115,21 @@ def recursive_cdu_heads(self, sloppy=False):
...
+            heads[c] = hd
+            return hd
+
         for c in self.cdus():
             get_head(c)
-        return cache
+        return heads
 
-    def without_cdus(self, sloppy=False):
+    def without_cdus(self, sloppy=False, mode='head'):
```
Comment (moreymat) [inline educe/stac/graph.py]: Could I bother you into improving the existing docstring while you are at it, i.e. expand the docstring for this function to document both the existing `sloppy` and newly introduced `mode` keyword arguments?
e.g.
```
Parameters
---------------
sloppy: boolean, default=False
    blablabla

mode: string, default='head'
    Replacement strategy for edges related to CDUs. "head" blablabla, "broadcast" blablabla, "custom" blablabla.
"""

**Comment by moreymat** [comment]
@jrmyp I am done with this round of reviewing, all these should be very quick fixes (mostly cosmetic and documentation).
Please ping me when you are done, so we can iterate fast and try to merge this PR shortly.

File: educe/graph.py (hunk 1 above)
Comment (jrmyp) [inline educe/graph.py]: That's... an excellent question, I don't know why I had a list there. Fixed.

File: educe/graph.py (hunk 2 above)
Comment (jrmyp) [inline educe/graph.py]: Fixed. Why not `self.cdu_members(m, deep=True)`, by the way ?

File: educe/stac/graph.py (hunk 3 above)
Comment (jrmyp) [inline educe/stac/graph.py]: Fixed.

File: educe/stac/graph.py (hunk 4 above)
Comment (jrmyp) [inline educe/stac/graph.py]: Fixed.

File: educe/stac/graph.py (hunk 5 above)
Comment (jrmyp) [inline educe/stac/graph.py]: I prefer it too, it's cleaner and more readable. Fixed.

File: educe/stac/graph.py (hunk 6 above)
Comment (jrmyp) [inline educe/stac/graph.py]: Fixed.

**Comment by jrmyp** [comment]
The cosmetic fixes are done, doc is in progress. I still need to test the annotation rewriting (to replace completely the original `strip_cdus` method). I should be done in a few hours (less, if there are no bugs).

File: educe/graph.py (hunk 2 above)
Comment (moreymat) [inline educe/graph.py]: Makes absolutely explicit that the value of `deep` is carried from the current state to recursive calls. Mostly a hunch, though.

**Comment by moreymat** [comment]
Good, let us try to keep having short iterations wherever possible.

File: educe/graph.py (hunk 1 above)
Comment (jrmyp) [inline educe/graph.py]: Now I remember, after it crashes: you can't add elements to a set while you are iterating over it. Hence my copy of `members` into a list. I could have copied it into a new set, though...

**Comment by jrmyp** [comment]
@moreymat: The code passed my tests, running on the tiny corpus. I now need to run all the pipeline with different extracting options.

Regarding feature extraction, I'm referencing irit-melodi/irit-stac#16 to enable the pipeline to use the change.

I saw the updated statistics PR, and I'll try using it with the new graphs.

File: educe/graph.py (hunk 1 above)
Comment (moreymat) [inline educe/graph.py]: Good catch. A list is fine, then. LGTM.

**Comment by moreymat** [comment]
LGTM. When you think this is ready to merge, please prepend [MRG + 1] to the title and merge.

**Pull request merged**